motor/providers/gcpinstancesnapshot/snapshot.go

```diff
@@ -97,28 +98,92 @@ func (sc *SnapshotCreator) InstanceInfo(projectID, zone, instanceName string) (i
 	}
 
 	if bootDisk != nil {
-		ii.BootDiskSource = bootDisk.Source
+		ii.BootDiskSourceURL = bootDisk.Source
 	}
 
 	return ii, nil
 }
 
-// cloneDisk clones a provided disk
-func (sc *SnapshotCreator) cloneDisk(sourceDisk, projectID, zone, diskName string) (string, error) {
-	var clonedDiskUrl string
+type snapshotInfo struct {
+	PlatformMrn  string
+	ProjectID    string
+	SnapshotName string
+	SnapshotUrl  string
+}
+
+func (sc *SnapshotCreator) SnapshotInfo(projectID, snapshotName string) (snapshotInfo, error) {
+	ctx := context.Background()
+	si := snapshotInfo{}
+
+	computeService, err := sc.computeServiceClient(ctx)
+	if err != nil {
+		return si, err
+	}
+
+	snapshot, err := computeService.Snapshots.Get(projectID, snapshotName).Context(ctx).Do()
+	if err != nil {
+		return si, err
+	}
+
+	si.ProjectID = projectID
+	si.SnapshotName = snapshot.Name
+	si.SnapshotUrl = snapshot.SelfLink
+	si.PlatformMrn = SnapshotPlatformMrn(projectID, snapshot.Name)
+
+	return si, nil
+}
+
+// searchLatestSnapshot looks for the latest available snapshot for the instance
+func (sc *SnapshotCreator) searchLatestSnapshot(projectID, sourceDiskUrl string) (string, error) {
 	ctx := context.Background()
 
 	computeService, err := sc.computeServiceClient(ctx)
 	if err != nil {
 		return "", err
 	}
 
-	// create a new disk clone
-	disk := &compute.Disk{
-		Name:       diskName,
-		SourceDisk: sourceDisk,
-		Labels:     sc.labels,
+	latestSnapshotTimestamp := time.UnixMilli(0)
+	var latestSnapshot *compute.Snapshot
+
+	req := computeService.Snapshots.List(projectID)
+	if err := req.Pages(ctx, func(page *compute.SnapshotList) error {
+		for _, snapshot := range page.Items {
+			// we are only interested in disks that are attached to the
+			if snapshot.SourceDisk != sourceDiskUrl {
+				continue
+			}
+
+			// RFC3339 encoded like 2021-02-28T02:31:38.654-08:00
+			snapshotCreated, err := time.Parse(time.RFC3339, snapshot.CreationTimestamp)
+			if err != nil {
+				log.Err(err).Str("snapshot", snapshot.Name).Str("creation-timestamp", snapshot.CreationTimestamp).Msg("snapshot timestamp is not parsable")
+				// we ignore snapshots that we cannot parse
+				continue
+			}
+
+			if latestSnapshotTimestamp.Before(snapshotCreated) {
+				latestSnapshot = snapshot
+				latestSnapshotTimestamp = snapshotCreated
+			}
+		}
+		return nil
+	}); err != nil {
+		return "", err
+	}
+
+	return latestSnapshot.SelfLink, nil
```

Comment:
This panics, when there is no snapshot:
```
cnquery scan gcp instance cz-test-cnquery-pr2 --zone europe-west1-b --project-id mondoo-dev-262313 --use-latest-snapshot --verbose
...
→ discover related assets for 1 asset(s)
DBG run resolver resolver="GCP Compute Instance Resolver" resolver-id=gcp-compute-instance
→ resolved assets resolved-assets=1
DBG connecting to asset cz-test-cnquery-pr2
DBG establish connection to asset connection=gcp-compute-instance insecure=false
DBG establish motor connection
...
DBG platform> detected os family=["debian","linux","unix","os"] platform=debian
DBG local> run command curl --noproxy '*' -H Metadata-Flavor:Google http://metadata.google.internal/computeMetadata/v1/project/project-id?alt=json
DBG local> run command curl --noproxy '*' -H Metadata-Flavor:Google http://metadata.google.internal/computeMetadata/v1/instance/id?alt=json
DBG local> run command curl --noproxy '*' -H Metadata-Flavor:Google http://metadata.google.internal/computeMetadata/v1/instance/name?alt=json
DBG local> run command curl --noproxy '*' -H Metadata-Flavor:Google http://metadata.google.internal/computeMetadata/v1/instance/zone?alt=json
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xe8 pc=0x80ebfce]

goroutine 35 [running]:
go.mondoo.com/cnquery/motor/providers/gcpinstancesnapshot.(*SnapshotCreator).searchLatestSnapshot(0x8fa4360?, {0x7ffd0a73b75e, 0x11}, {0xc000a84a10, 0x6f})
	/home/christian/workspace/mondoo/github.com/cnquery/motor/providers/gcpinstancesnapshot/snapshot.go:174 +0x1ce
go.mondoo.com/cnquery/motor/providers/gcpinstancesnapshot.New(0xc00059fc80)
	/home/christian/workspace/mondoo/github.com/cnquery/motor/providers/gcpinstancesnapshot/provider.go:132 +0x75e
...
```

With the above command, I tried to scan an instance which has no snapshots.

Reply:
Great catch, will adjust the code.